metagraph/src/graph/representation/succinct/boss_chunk_construct.cpp

```diff
@@ -673,114 +710,225 @@ void recover_dummy_nodes_disk(KmerCollector &kmer_collector,
+    using T_INT = get_int_t<T>;
+    using KMER = get_first_type_t<T>; // 64/128/256-bit KmerBOSS with sentinel $
+    using KMER_INT = typename KMER::WordType; // 64/128/256-bit integer
+
+    auto kmers_out = std::make_shared<ChunkedWaitQueue<T_INT>>(ENCODER_BUFFER_SIZE);
+
+    if (is_first_chunk) {
+        // add the main dummy source k-mer
+        if constexpr(utils::is_pair_v<T>) {
+            kmers_out->push({ KMER_INT(0), 0 });
+        } else {
+            kmers_out->push(KMER_INT(0));
+        }
     }
 
     // push all other dummy and non-dummy k-mers to |kmers_out|
-    async_worker.enqueue([k, kmers_out, real_split_by_W, dummy_chunks]() {
-        const KMER_INT kmer_delta
-                = kmer::get_sentinel_delta<KMER_INT>(KMER::kBitsPerChar, k + 1);
+    ThreadPool async_worker(1, 1);
+    async_worker.enqueue([kmers_out, real_name, dummy_names]() {
+        Decoder<T_INT> decoder(real_name, true);
 
-        common::Transformed<common::MergeDecoder<T_INT_REAL>, T> decoder(
-            [&](const T_INT_REAL &v) {
-                if constexpr(utils::is_pair_v<T>) {
-                    return T(kmer::transform<KMER>(reinterpret_cast<const KMER_REAL &>(v.first), k + 1)
-                                + kmer_delta,
-                             v.second);
-                } else {
-                    return kmer::transform<KMER>(reinterpret_cast<const KMER_REAL &>(v), k + 1) + kmer_delta;
-                }
-            },
-            real_split_by_W, true
-        );
-
-        common::Transformed<common::MergeDecoder<KMER_INT>, T> decoder_dummy(
+        common::Transformed<common::MergeDecoder<KMER_INT>, T_INT> decoder_dummy(
             [](const KMER_INT &v) {
                 if constexpr(utils::is_pair_v<T>) {
-                    return T(reinterpret_cast<const KMER &>(v), 0);
+                    return std::make_pair(v, 0);
                 } else {
-                    return reinterpret_cast<const KMER &>(v);
+                    return v;
                 }
             },
-            dummy_chunks, true
+            dummy_names, true
         );
 
-        while (!decoder.empty() && !decoder_dummy.empty()) {
-            if (get_first(decoder.top()) < get_first(decoder_dummy.top())) {
-                kmers_out->push(decoder.pop());
+        std::optional<T_INT> next = decoder.next();
+        while (next && !decoder_dummy.empty()) {
+            if (get_first(next.value()) < get_first(decoder_dummy.top())) {
+                kmers_out->push(next.value());
+                next = decoder.next();
             } else {
                 kmers_out->push(decoder_dummy.pop());
             }
         }
-        while (!decoder.empty()) {
-            kmers_out->push(decoder.pop());
+        while (next) {
+            kmers_out->push(next.value());
+            next = decoder.next();
         }
         while (!decoder_dummy.empty()) {
             kmers_out->push(decoder_dummy.pop());
```

Comment:
This is optional, but do you think this can be made simpler with `std::merge` if the right interface is there?

Reply:
Yeah, would be good, but it requires iterators and decoders have a different interface